### ISSUE TITLE
SSLHandler slows down large file transfer

### ISSUE THREAD

**Issue by konsultaner** [netty/netty#6139]
I try to send a file larger then 5MB over the following pipeline:

1. SSLHandler
2. ByteToMessageCodec
3. IdleStateHandler
4. SimpleChannelInboundHandler

Sending a 10MB file takes arround 3Sekonds. When I remove the SSLHandler the file is transferred within 800ms. I found out that the SSLHandler Chunks the bytebuffer into very tiny pieces, that take forever to be sent.

I use the `UnpooledByteBufAllocator.DEFAULT` in my ChannelOptions. my encode method form the `ByteToMessageCodec` looks like this:

```java
protected void encode(ChannelHandlerContext channelHandlerContext, IMessage iMessage, ByteBuf byteBuf) throws Exception {
    if(byteBuf.isWritable()){
        byteBuf.writeBytes((serializer.serialize(iMessage) + RECORD_SEPARATOR).getBytes());
    }
}
```
Is there a way to tweek the SSLHandler?

Maybe the SSLEngine is the troublemaker? I found the following line of code inside the SSLHandler, but this constructor is depricated:

```java
wantsLargeOutboundNetworkBuffer = !opensslEngine;
```

### Netty version
4.0.40

### JVM version
java version "1.8.0_111"
Java(TM) SE Runtime Environment (build 1.8.0_111-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.111-b14, mixed mode)

### OS version
Win10

### StackOverflow
[netty-sslhandler-slows-down-file-transfer-5mb](http://stackoverflow.com/questions/41182586/netty-sslhandler-slows-down-file-transfer-5mb)

**Comment by Scottmitch** [comment]
Thanks for reaching out @konsultaner. Please provide a minimal reproducer so we can take a closer look.

**Comment by konsultaner** [comment]
@Scottmitch Do you want me to write a demo project? Or just more of my code?

**Comment by johnou** [comment]
@konsultaner demo project would be excellent.

**Comment by normanmaurer** [comment]
@konsultaner can you also please test with latest code from 4.1 branch ?

**Comment by konsultaner** [comment]
@normanmaurer Ok I'll test that as well. I already build the usecase but I can't get it running, maybe you can take a look at it. It does not send the file to the receiver. I don't know what I'm doing wrong in the example code - embarrassing :see_no_evil:
[ssldemo.zip](https://github.com/netty/netty/files/659446/ssldemo.zip)

**Comment by Scottmitch** [comment]
@konsultaner - Why do you need to create new threads? Consider starting with [the echo example](https://github.com/netty/netty/tree/4.1/example/src/main/java/io/netty/example/echo) as a template and provide your reproducer on top of this.

**Comment by konsultaner** [comment]
@Scottmitch thank you for the hint. I also figured out what I did wrong. I attatched a demo based on the echo server. @normanmaurer It also appears in netty 4.1.6

The demo contains the exact stack that I used in my original stack. To swap from SSL to non-SSL change ´EchoClient.SSL´ the pom.xml is currently set to netty 4.1.6. you may change it to 4.0.x and see the same result. I added a 10MB File. Sending it with SSL costs 1Sekond and 200ms without - 5x faster.
[ssldemoEcho.zip](https://github.com/netty/netty/files/659613/ssldemoEcho.zip)

**Comment by Scottmitch** [comment]
I have not looked at the code yet ... but have you tried [netty-tcnative](http://netty.io/wiki/forked-tomcat-native.html#how-to-download-netty-tcnative-boringssl-static)? It directly uses OpenSSL which generally is more performant than JDK's SSL.

**Comment by konsultaner** [comment]
@Scottmitch I took a look a netty-tcnative which sound pretty interesting, but it seems like it is not platform indepandent or the VM needs extra sources. This is something I do not really want in my project. I'm happy to have a out of the box solution.

I will give it a try anyway to optionally provide extra performance.

Seems like you think this is an expected behaviour?

**Comment by Scottmitch** [comment]
> but it seems like it is not platform indepandent or the VM needs extra sources.

Yes the JAR contains a JNI library which varies for different platforms. `netty-tcnative-boringssl-static` includes libraries pre-built for the "most used" platforms (e.g. windows/osx/linux).

> Seems like you think this is an expected behaviour?

It is generally expected that using TLS/SSL will incur a performance penalty vs no TLS/SSL. It is generally expected that the penalty will be less when using `netty-tcnative` vs JDK. How much slower depends upon the use case, configuration, and potentially hardware support for crypto. I still recommend trying `netty-tcnative` to get an idea of performance differences since that is what this issue is about.

**Comment by konsultaner** [comment]
> It is generally expected that using TLS/SSL will incur a performance penalty vs no TLS/SSL

@Scottmitch @normanmaurer If this is the case, you may as well close this issue. A little hint in the docs would probably help others, because 1/5 of the performance was quite unexpected to me. Thank you for the netty-tcnative hint. I will give it a try 👍.

**Comment by fredericBregier** [comment]
From my own experience, 1/5 is quite huge. Generally I observed 10 to 20% of reduction of bandwidth , but with high cpu usage using standard java implementation, not jni ones.

**Comment by konsultaner** [comment]
@fredericBregier Thats what I thought! Thats why I added a [demo](https://github.com/netty/netty/files/659613/ssldemoEcho.zip). Maybe I did something completely wrong. Would be nice if there was a solution to it.

**Comment by Scottmitch** [comment]
@konsultaner - I took a quick look at your example. Here is a general tip:

- The SSL engines generally need a direct ByteBuffer to do encryption. So they will copy to a direct byte buffer if your data is not in a direct byte buffer. In your example you parse the data into a `byte[]` (`fileInputStreamReader.read(bytes);`),  then copy this data into a `String` (`new String(bytes, StandardCharsets.UTF_8))`), then translate back to a `byte[]` (`msg.getBytes()`) and copy this to a `ByteBuf`. Try allocating your file into a direct ByteBuf up front and just reusing this data instead of doing all the intermediate transformations and copies.

Some questions:
- `EchoClientHandler#encode` checks if the output buffer is writable before writing to it. Can you explain this? Why are you using `ByteToMessageCodec` in the first place? You should be able to just directly write a direct ByteBuf to the channel.
- `EchoClientHandler#decode` wait for the entire file to be buffered in memory, and then converts to a String. Do you have to wait for the entire file, and do you have to convert to a `String`?

**Comment by fredericBregier** [comment]
@konsultaner In addition to the very good points from @Scottmitch I would suggest you to use the stream support from Netty, since your title is containing "large file", which means for me "not able to stay in memory" (either as a byte array, string or whatever buffer).

See https://github.com/netty/netty/tree/4.1/handler/src/main/java/io/netty/handler/stream

Doing this way will prevent to load file into bytes then string then back to bytes to buffer... therefore preventing multiple copies. Another way would be to "reuse" the very same buffer, reading the file by chunk, manually.

**Comment by konsultaner** [comment]
@Scottmitch Thank you for those usefull tips! This will improve my software a lot!

>EchoClientHandler#encode checks if the output buffer is writable before writing to it. Can you explain this?

Actually not, I think I saw it somewhere an added it to prevent errory that I didn't yet know about. Is it not needed?

>Why are you using ByteToMessageCodec in the first place? You should be able to just directly write a direct ByteBuf to the channel.

I use it because this is just a code that reproduces my problem that I have with a more complex project.

>EchoClientHandler#decode wait for the entire file to be buffered in memory, and then converts to a String. Do you have to wait for the entire file, and do you have to convert to a String?

I assume you want to hint me to the streaming possibilies that netty offers and that @fredericBregier mentioned. 

In general its worth knowing that I wrote a message broker and one of my potential clients told me about the big file issue. I would like to send tiny 500byte messages (most of the times) as well as 500MB messages as fast as possible. I will refactor the whole messaging to work with streaming or at least direct buffer usage. The conversion to string is a huge bottleneck, I know. 

There is just one question left for me. Do I need to take care of the memory allocation and deallocation by myself. @normanmaurer  mentioned that netty 4.1.x now uses pooled direct buffers by default because memory leak monitoring is super easy now? I havn't found a proper source about buffers in netty yet.

**Comment by Scottmitch** [comment]
> Actually not, I think I saw it somewhere an added it to prevent errory that I didn't yet know about. Is it not needed?

What is your fallback if the buffer is not writable? Currently the data is dropped, is this what you want? Also the buffer may be able to expand to accommodate the additional data (see `ByteBuf#ensureWritable`).

> I use it because this is just a code that reproduces my problem that I have with a more complex project.

Understood. If this same pattern exists in the other project you may want to reconsider how this case is handled there too.

> I assume you want to hint me to the streaming possibilies that netty offers

In general as files get larger then loading them completely into memory at once may limiting/impossible. If you have to load these files "on demand" (not a fixed set of files known up front) you may get more responsiveness from loading/sending them in chunks rather than all at once.

> The conversion to string is a huge bottleneck, I know.

In general avoid translating into `String` unless it is necessary. If it is necessary as your self why and if you can change APIs (or add new ones) to avoid this translation.

> Do I need to take care of the memory allocation and deallocation by myself

Depends what you mean by this. Netty provides a level of indirection (`ByteBufAllocator`) which allocates `ByteBuf` objects. We have a pooled implementation of the allocator (`PooledByteBufAllocator`) designed to help with GC pressure. Part of this implementation relies upon reference counting to ensure memory is returned back to the pool in a timely/efficient manner. See http://netty.io/wiki/reference-counted-objects.html for more details.

> ... netty 4.1.x now uses pooled direct buffers by default because memory leak monitoring is super easy now

I wouldn't say "super easy", but it depends upon your experience and familiarity with memory management. Understandably this can be a pain point for Java developers. We have also fixed some bugs related to memory leak detection which should be included in the next Netty release.
See http://netty.io/wiki/reference-counted-objects.html#troubleshooting-buffer-leaks for more details.

**Comment by konsultaner** [comment]
I know this is probably not the right place to ask question, but this issue seems to become something like a knowledge base for bytebuffers and SSL usage in netty ;) I have some further questions about zero copy that I placed on stack overflow. I would really appreciate an answer if possible.

http://stackoverflow.com/questions/41978934/how-do-i-correctly-work-with-bytetomessagecodec-encode-and-direct-buffers

**Comment by konsultaner** [comment]
I added the [ssl demo](https://github.com/netty/netty/files/659613/ssldemoEcho.zip) to demonstrate a slow file sending. In my project I finally found the major bottle neck. For some reason in the ssl demo the [channel.write](https://netty.io/4.1/xref/io/netty/channel/socket/nio/NioSocketChannel.html#412) sends the whole file in 4 iterations. For some reason my project sends the buffer in much more steps. In both cases the buffersize is the same. The buffer type is also the same. The [nioBuffer](https://netty.io/4.1/xref/io/netty/channel/socket/nio/NioSocketChannel.html#410) has only one entry. How can I influence the "package size" of written chunks per iteration?

**Comment by Scottmitch** [comment]
Are you sending to the same peer in both examples? How much data is accepted by the write to the OS is dependent on the local send buffer size, the peer's receive buffer size, and how fast the peer is reading data.

**Comment by konsultaner** [comment]
No the slow example is another programm. I found out that the channel write has become faster after changing the buffer size but the [channel read](https://netty.io/4.1/xref/io/netty/channel/socket/nio/NioSocketChannel.html#364) hangs in a loop at:

```java
...
protected int doReadBytes(ByteBuf byteBuf) throws Exception {
    Handle allocHandle = this.unsafe().recvBufAllocHandle();
    allocHandle.attemptedBytesRead(byteBuf.writableBytes());
    return byteBuf.writeBytes(this.javaChannel(), allocHandle.attemptedBytesRead());
}
...
```

This is called for 20 seconds and comes from the [NioEventLoop line 642](https://netty.io/4.1/xref/io/netty/channel/nio/NioEventLoop.html#642)

Sending seems therefor fine but receiving takes for ever.

**Comment by konsultaner** [comment]
I went deep down into the socket classes of java to find the issue about the read buffer. I had the channel options set to 50MB receive buffer. At the lowest java level where the [FileDescriptor](https://docs.oracle.com/javase/7/docs/api/java/io/FileDescriptor.html) is writing data to the bytebuffer. Java has allocated a bytebuffer with 50MB of size and wants to read just a bit less then that. But the result is that the buffer is only filled with about 12kb of data per read cycle. 

I have changed every channel option i found, but I have no Idea where I my tweak anything else. I could provide demo jars via email only, because it is a commercial project. I have implemented all the tips you gave me in this thread. Thanks a lot so far. This was a huge helper!!!

@Scottmitch @normanmaurer I would pay for some help. This issue gives me huge headaches and I have absolutly no idea how to solve it.

**Comment by Scottmitch** [comment]
Hey @konsultaner sorry this is still plaguing you. Unfortunately I don't have time to dig in ATM. I'm not sure if the ssldemo above is already reproducing this issue and minimal (only includes what is required to reproduce), but if you could provide this perhaps someone from the community can take a peek.

**Comment by normanmaurer** [comment]
@konsultaner actually @Scottmitch did a lot of changes related to this and I think it should be fixed in the latest netty release (4.1.18.Final). Let me close this and please reopen if you still think there is an issue
